### ISSUE TITLE
Cabal-3 breaks GHC's cabal01 test

### ISSUE THREAD

**Issue by bgamari** [haskell/cabal#6068]
GHC's `cabal01` test has a trivial `Setup.lhs` file:
```haskell
> import Distribution.Simple
> main = defaultMain
```

Bumping Cabal to 15675844bb36929448c189d6b4aabf7e853b3ee1 breaks this test:
```
=====> cabal01(normal) 1 of 3 [0, 0, 0]
cd "cabal/cabal01/cabal01.run" && $MAKE -s --no-print-directory cabal01 VANILLA=--enable-library-vanilla PROF=--disable-library-profiling DYN=--enable-shared CLEANUP=1  
Wrong exit code for cabal01()(expected 0 , actual 2 )
Stdout ( cabal01 ):
install1:
bin
lib
install2:
bin
lib
Stderr ( cabal01 ):
fromFlag NoFlag. Use fromFlagOrDefault
CallStack (from HasCallStack):
  error, called at libraries/Cabal/Cabal/Distribution/Simple/Flag.hs:96:21 in Cabal-3.0.0.0:Distribution.Simple.Flag
  fromFlag, called at libraries/Cabal/Cabal/Distribution/Simple/SrcDist.hs:126:20 in Cabal-3.0.0.0:Distribution.Simple.SrcDist
  sdist, called at libraries/Cabal/Cabal/Distribution/Simple.hs:412:5 in Cabal-3.0.0.0:Distribution.Simple
  sdistAction, called at libraries/Cabal/Cabal/Distribution/Simple.hs:188:50 in Cabal-3.0.0.0:Distribution.Simple
  defaultMainHelper, called at libraries/Cabal/Cabal/Distribution/Simple.hs:116:27 in Cabal-3.0.0.0:Distribution.Simple
  defaultMain, called at Setup.lhs:2:10 in main:Main
make[2]: *** [Makefile:35: cabal01] Error 1
*** unexpected failure for cabal01(normal)
```

**Comment by bgamari** [comment]
For the record the `Setup configure` command line is 
```
./setup configure -v0 --prefix=/opt/exp/ghc/ghc-8.8/testsuite/tests/cabal/cabal01/cabal01.run/install --with-compiler=/opt/exp/ghc/ghc-8.8/_build/stage1/bin/ghc --ghc-options=-dcore-lint -dstg-lint -dcmm-lint -no-user-package-db -fno-warn-missed-specialisations -fshow-warning-groups -fdiagnostics-color=never -fno-diagnostics-show-caret -Werror=compat -dno-debug-output --with-hc-pkg=/opt/exp/ghc/ghc-8.8/_build/stage1/bin/ghc-pkg --package-db=local.db --enable-library-vanilla --enable-library-profiling --enable-shared --libsubdir=$pkgid
```

**Comment by bgamari** [comment]
The problem is this`setup sdist` invocation:
```
Distribution quality errors:
No 'build-type' specified. If you do not need a custom Setup.hs or ./configure
script then use 'build-type: Simple'.
No 'synopsis' or 'description' field.
The 'license' field is missing or is NONE.
Distribution quality warnings:
No 'category' field.
No 'maintainer' field.
It is recommended that the 'cabal-version' field only specify a version range
of the form '>= x.y'. Use 'cabal-version: >= 0'. Tools based on Cabal 1.10 and
later will ignore upper bounds.
From version 1.24 cabal supports specifying explicit dependencies for Custom
setup scripts. Consider using cabal-version >= 1.24 and adding a
'custom-setup' section with a 'setup-depends' field that specifies the
dependencies of the Setup.hs script itself. The 'setup-depends' field uses the
same syntax as 'build-depends', so a simple example would be 'setup-depends:
base, Cabal'.
Note: the public hackage server would reject this package.
fromFlag NoFlag. Use fromFlagOrDefault
CallStack (from HasCallStack):
  error, called at libraries/Cabal/Cabal/Distribution/Simple/Flag.hs:96:21 in Cabal-3.0.0.0:Distribution.Simple.Flag
  fromFlag, called at libraries/Cabal/Cabal/Distribution/Simple/SrcDist.hs:126:20 in Cabal-3.0.0.0:Distribution.Simple.SrcDist
  sdist, called at libraries/Cabal/Cabal/Distribution/Simple.hs:412:5 in Cabal-3.0.0.0:Distribution.Simple
  sdistAction, called at libraries/Cabal/Cabal/Distribution/Simple.hs:188:50 in Cabal-3.0.0.0:Distribution.Simple
  defaultMainHelper, called at libraries/Cabal/Cabal/Distribution/Simple.hs:116:27 in Cabal-3.0.0.0:Distribution.Simple
  defaultMain, called at Setup.lhs:2:10 in main:Main
```

**Comment by bgamari** [comment]
I believe this should be fixed by #6073.

**Comment by bgamari** [comment]
This was resolved by #6073.
